Approximate gap Jacobians in GapConductanceConstraint

lindsayad opened the pull request:
Here's the difference between user Jac and hand-coded Jac for the Lagrange multiplier degrees of freedom

Without displacements coupled in:
```
row 100: (1, 5.37765)  (7, 1.68051)  (49, -4.70544)  (51, -2.35272)
row 101: (1, 4.39623)  (7, 9.35711)  (15, 0.362991)  (49, -2.35272)  (51, -9.41087)  (59, -2.35272)
row 102: (1, 0.107553)  (7, 7.86481)  (15, 6.13052)  (21, 0.0134441)  (51, -2.35272)  (59, -9.41087)  (67, -2.35272)
row 103: (7, 0.860423)  (15, 10.3789)  (21, 2.87704)  (59, -2.35272)  (67, -9.41087)  (75, -2.35272)
row 104: (15, 2.87704)  (21, 10.3789)  (27, 0.860423)  (67, -2.35272)  (75, -9.41087)  (83, -2.35272)
row 105: (15, 0.0134441)  (21, 6.13052)  (27, 7.86481)  (33, 0.107553)  (75, -2.35272)  (83, -9.41087)  (91, -2.35272)
row 106: (21, 0.362991)  (27, 9.35711)  (33, 4.39623)  (83, -2.35272)  (91, -9.41087)  (99, -2.35272)
row 107: (27, 1.68051)  (33, 5.37765)  (91, -2.35272)  (99, -4.70544)
```
With displacements coupled in:
```
row 100:
row 101:
row 102:
row 103:
row 104:
row 105:
row 106:
row 107:
```
there are still missing Jacobians for the temperature degrees of freedom corresponding to the fact that we don't know how to compute the dependence of JxW and test functions on the displacement dofs, but the fix for that to me continues to be AD in libmesh geometric quantities.

Reasons on why I think this PR should help thermal contact convergence
- This PR doesn't seem to affect the overall Jacobian ratio very much...if displacements are not coupled into the mortar object, the Jacobian ratio is 5e-3. If they are coupled in, the ratio only drops to 3e-3. However, the Jacobian ratio for the lagrange multipler matrix sub-block should be vastly improved. (It actually appears to be perfect when viewing the final nonlinear iteration)
- I also switched the sign on the lagrange multiplier residual, which means the diagonal should be positive instead of negative now, which should be better for a linear solver.
- Instead of taking the norm of the gap vector, computed by differencing the primary and secondary points, we obtain a scalar value now by dotting with the normal...which is also way more representative of the physics we are trying to solve here! Before you could have had a non-zero gap even when the gap is closed if the primary and secondary points were offset in the tangential direction (!)
- Finally, preventing that heat flux from going to infinity as the gap closes can only be a good thing right? Even if we're making the function discontinuous with that `std::max`

Closes #13488

dschwen committed:
Approximate gap derivatives (#13488)

lindsayad committed:
Make lm diagonal entry positive instead of negative

lindsayad committed:
Add min_gap parameter to GapConductanceConstraint

lindsayad committed:
Add tests with production tensor mechanics objects plus mechanical contact (#13488)

Complete test matrix:

2x2x2x2 [low-k, high-k] x [first order, second order] x [small,finite]
          x [cartesian, RZ]

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/700819/) on c5d06f7 wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/17252/site)

This comment will be updated on new commits.

lindsayad committed:
Protect against deriv insertion for node without active displacements

lindsayad commented:
Actually looking pretty good!
Without displacements coupled in:
```
row 100: (1, 5.37765)  (7, 1.68051)  (49, -4.70544)  (51, -2.35272)
row 101: (1, 4.39623)  (7, 9.35711)  (15, 0.362991)  (49, -2.35272)  (51, -9.41087)  (59, -2.35272)
row 102: (1, 0.107553)  (7, 7.86481)  (15, 6.13052)  (21, 0.0134441)  (51, -2.35272)  (59, -9.41087)  (67, -2.35272)
row 103: (7, 0.860423)  (15, 10.3789)  (21, 2.87704)  (59, -2.35272)  (67, -9.41087)  (75, -2.35272)
row 104: (15, 2.87704)  (21, 10.3789)  (27, 0.860423)  (67, -2.35272)  (75, -9.41087)  (83, -2.35272)
row 105: (15, 0.0134441)  (21, 6.13052)  (27, 7.86481)  (33, 0.107553)  (75, -2.35272)  (83, -9.41087)  (91, -2.35272)
row 106: (21, 0.362991)  (27, 9.35711)  (33, 4.39623)  (83, -2.35272)  (91, -9.41087)  (99, -2.35272)
row 107: (27, 1.68051)  (33, 5.37765)  (91, -2.35272)  (99, -4.70544)
```
With displacements coupled in:
```
row 100:
row 101:
row 102:
row 103:
row 104:
row 105:
row 106:
row 107:
```

tophmatthews commented on modules/combined/test/tests/gap_heat_transfer_mortar/gap_heat_transfer_mortar_displaced.i:
```diff
@@ -25,33 +25,18 @@
 []
 
-[AuxVariables]
+[Variables]
```
How about a test with reference residual? Maybe you need #17249 first...

lindsayad replied:
Yes I would

recuero replied:
Second order meshes seem to show more convergence challenges for the cases I have tried.

lindsayad commented:
@tophmatthews can you try this out on whatever problem you were doing yesterday? Reasons on why I think this PR should help
- This PR doesn't seem to affect the overall Jacobian ratio very much...if displacements are not coupled into the mortar object, the Jacobian ratio is 5e-3. If they are coupled in, the ratio only drops to 3e-3. However, the Jacobian ratio for the lagrange multipler matrix sub-block should be vastly improved. 
- I also switched the sign on the lagrange multiplier residual, which means the diagonal should be positive instead of negative now, which should be better for a linear solver.
- Instead of taking the norm of the gap vector, computed by differencing the primary and secondary points, we obtain a scalar value now by dotting with the normal...which is also way more representative of the physics we are trying to solve here! Before you could have had a non-zero gap even when the gap is closed if the primary and secondary points were offset in the tangential direction (!)
- Finally, preventing that heat flux from going to infinity as the gap closes can only be a good thing right? Even if we're making the function discontinuous with that `std::max`

lindsayad commented on modules/combined/test/tests/gap_heat_transfer_mortar/gap_heat_transfer_mortar_displaced.i:
```diff
@@ -82,17 +67,27 @@
   [../]
+  [disp_x]
+    type = Diffusion
+    variable = disp_x
```
Ok in the most recent commit I've adapted one of your `mortar_tm` bouncing block contact tests to include `GapConductanceConstraint` and heat conduction. Some interesting stats from that input:

Newton:

135 time steps
538 total nls
16.883 seconds
average 4.0 nls/timestep

PJFNK:

135 time steps
415 total nls
21.029 seconds
average 3.1 nls/timestep

tophmatthews replied:
What about compared to nonmortar?

tophmatthews commented on modules/heat_conduction/src/constraints/GapConductanceConstraint.C:
```diff
@@ -39,9 +56,23 @@ GapConductanceConstraint::computeQpResidual(Moose::MortarType mortar_type)
+
+      auto l =
+          std::max((ad_phys_points_primary - ad_phys_points_secondary) * _normals[_qp], _min_gap);
+      return (_lambda[_qp] - _k * (_u_primary[_qp] - _u_secondary[_qp]) / l) * _test[_i][_qp];
```
And high k...probably need to test with k = 1 and k = 100? Or maybe it won't matter?

tophmatthews replied:
I mean, push back if you want, I just know that cases like that in which are quick for you to test now can really head off confusion down the line

lindsayad replied:
That'll definitely make the scaling worse and the problem harder to solve...which seems to be what you want haha

tophmatthews replied:
I don't think it'll ever go above 100, usually about 60. May go less than 1 when doing a gas gap, I forget exactly.

recuero commented on framework/src/base/Assembly.C:
```diff
@@ -944,14 +944,15 @@ Assembly::computeAffineMapAD(const Elem * elem,
         if (_computing_jacobian)
           for (const auto & disp_num : _displacements)
-            Moose::derivInsert(elem_point(dimension++).derivatives(),
+            if (node.n_dofs(sys_num, disp_num))
```
As far as I understand this was possibly doing some unnecessary work before this PR?

lindsayad replied:
More insidious than that...this would put derivatives into these geometric quantities even if the displacement variables were auxiliary variables. So you can imagine that if the aux var displacements were numbered 0 and 1 in the aux system, and you had some nl variables, let's say T and u numbered 0 and 1 respectively in the nl system, you would actually be inserting geometric quantity dependence on T and u!

lindsayad replied:
Wait this check is actually still bad... we're still going to enter this block under the exact scenario I posted above anytime T and u have dofs on the node!

lindsayad replied:
Whew, actually things are ok because of this code:
```
void
DisplacedProblem::init()
{
  for (THREAD_ID tid = 0; tid < libMesh::n_threads(); ++tid)
  {
    _assembly[tid]->init(_mproblem.couplingMatrix());
    std::vector<unsigned> disp_numbers;
    for (const auto & disp_string : _displacements)
    {
      const auto & disp_variable = getVariable(tid, disp_string);
      if (disp_variable.kind() == Moose::VarKindType::VAR_NONLINEAR)
        disp_numbers.push_back(disp_variable.number());
    }
    _assembly[tid]->assignDisplacements(std::move(disp_numbers));
  }
```
I check to make sure the displacement variable is nonlinear before passing it to `Assembly`.

Ok so after that whirlwind: the real reason this check is necessary is in case you are at a place in the mesh where the displacement variable doesn't actually exist. If we don't do this `n_dofs` check here, then we would fail an assertion when we call `node.dof_number`

recuero commented on modules/heat_conduction/src/constraints/GapConductanceConstraint.C:
```diff
@@ -39,9 +56,23 @@ GapConductanceConstraint::computeQpResidual(Moose::MortarType mortar_type)
+      // an approximation of the true phys points derivatives
+      for (unsigned int i = 0; i < _n_disp; ++i)
+      {
+        ad_phys_points_primary(i).derivatives() = (*_disp_primary[i])[_qp].derivatives();
+        ad_phys_points_secondary(i).derivatives() = (*_disp_secondary[i])[_qp].derivatives();
```
We may be able to do something like this for LWR...

recuero replied:
Hmm I can't see that we are actually grasping derivative info from the displacement variables to create the distance variable. Members `_phys_points_primary` and `_phys_points_secondary` reside in `MortarConstraintBase` so I believe we should still do it explicitly for the LWR case. I can take a further look at that.

lindsayad replied:
Oh you are right!

recuero reviewed (commented):
👍

tophmatthews commented:
Well, this is a new one...

```
Time Step 1, time = 100, dt = 100

Performing automatic scaling calculation

 0 Nonlinear |R| = 1.580149e+00
    |residual|_2 of individual variables:
                     temp:   1.58015
                     lm:     1.55353e-12
                     disp_x: 1.90059e-08
                     disp_y: 7.91669e-10
*** stack smashing detected ***: /usr/projects/packages/mule/topher/test/bison/bison-opt terminated

===================================================================================
=   BAD TERMINATION OF ONE OF YOUR APPLICATION PROCESSES
=   PID 29897 RUNNING AT sn043
=   EXIT CODE: 11
=   CLEANING UP REMAINING PROCESSES
=   YOU CAN IGNORE THE BELOW CLEANUP MESSAGES
===================================================================================
YOUR APPLICATION TERMINATED WITH THE EXIT STRING: Segmentation fault (signal 11)
This typically refers to a problem with your application.
Please see the FAQ page for debugging suggestions
```

tophmatthews commented:
Well, the stack smashing was due to AD container size being too small. It's going now, and looks to be performing much better! We'll see if the run finishes. I feel like I'll need the reference residual fixes to really have it kick ass since I'm spending unnecessary electrons cranking down on the lagrange multiplier.

tophmatthews commented:
Well it gets closer, 1.87e6 seconds in 1985 nl, vs 1.68e6 in 1999 nl, but then the temperature fails to converge. I think there is something else going on with my model though I have to debug. This certainly helps.

lindsayad commented:
#17189 contains some changes to the way we build the mortar mesh that I believe increases robustness, perhaps especially for second order meshes. Hopefully we can get that merged soon and you can experience that as well.

tophmatthews commented:
Now I get to 2.23e6 with second order LM, second order Mesh, this and the other fix in...

lindsayad commented:
But it still fails

tophmatthews commented:
Yes...but I still think I'm missing something.

tophmatthews commented:
But it's all helping, so get it in!

srinath-chakravarthy commented:
this is great that this is included into the framework. Just a quick comment, i think the gap calculations should be embedded in a 
if (_has_primary) block. In my experience this resolves choice of primary/secondary surfaces. 

Cheers
Srinath

lindsayad commented on modules/combined/test/tests/gap_heat_transfer_mortar/gap_heat_transfer_mortar_displaced.i:
```diff
@@ -123,6 +119,28 @@
+  []
+  [bottom_disp_y]
+    type = DirichletBC
+    preset = false
```
This was copied over from @tophmatthews 's test. I will say that for non-zero Dirichlet values for displacements I really like to use non-preset because if you use preset then you can actually blow a Dirichlet node past a non-Dirichlet node and create inverted elements

lindsayad committed:
Follow Srinath's suggestion and encapsulate gap calcs

If we don't project onto a primary face then there should be no heat
transfer, e.g. our residual should just be lambda, e.g. we drive lambda
(representing the heat flux) to zero in regions of the interface where
we don't have an accompanying primary face

lindsayad committed:
Fix normal used for determining mortar element orientation

lindsayad committed:
Only use displacement derivatives on displaced mesh (#13488)

lindsayad committed:
Apply snesmf_reuse_base = false as indicated by devel failures

lindsayad committed:
Update default 'c' value for ApplyPenetrationLMMechanicalContact

lindsayad commented:
@tophmatthews as of right now I've got this 2x2 matrix:

[k = 1, k = 100] x [order = first, order = second]

for small strain tensor mechanics. I'm guessing you're going to ask for finite strain :laughing:

tophmatthews commented:
I mean...you could guess without me asking even. 2DRz too 😄 
At least I don't ask for 3D!

lindsayad commented:
Alright I think if this passes tests this is ready to go. Unless @tophmatthews disagrees

recuero commented on modules/contact/test/tests/mortar_tm/2d/frictionless_first/finite_noaction.i:
```diff
@@ -201,6 +201,7 @@ name = 'finite_noaction'
   dtmin = 0.1
   timestep_tolerance = 1e-6
   line_search = 'contact'
+  snesmf_reuse_base = false
```
I have not see much of a difference flipping this Boolean recently: Does using dual basis functions "cancel" its effect?

lindsayad replied:
If you are using our new mortar objects, you should not see any effect because they are not stateful (thank goodness), e.g. F(u) is always the same for a given u. The old non-weighted-gap way seems to have potentially been stateful, or some other object in these inputs were stateful such that F(u) is not always the same for a given u.

recuero replied:
Got it

tophmatthews replied:
Maybe there's already logic, but can you throw a paramError if this is set and using the new mortar objects? I have a big personal distaste for input parameters that don't always do something since some one much less capable (me) has to track down what they do.

lindsayad replied:
What do you mean by "don't always do something"? This parameter will always do something. It toggles on/off whether we reuse the nonlinear residual as the base linear residual for matrix-free methods. This defaults to true in MOOSE but users can set it to false if they have to...and most contact methods have to, just not this new mortar. Auto-toggling of this parameter is pretty easy to achieve from an action (it's automatically toggled to `false` in the `ContactAction`) but it would be somewhat ugly to try to pull the `Executioner` option into the mortar constraint...then you would be getting a `paramError` from a block (the constraint block) that is not logically associated with the block you are issuing a parameter error about

recuero commented on modules/combined/test/tests/gap_heat_transfer_mortar/finite-2d-rz/finite.i:
```diff
@@ -0,0 +1,356 @@
+  type = Transient
+  solve_type = 'PJFNK'
+  petsc_options = '-snes_converged_reason -ksp_converged_reason'
+  petsc_options_iname = '-pc_type -mat_mffd_err -pc_factor_shift_type -pc_factor_shift_amount -snes_max_it'
```
Are you reducing the default number of iterations with `-snes_max_it`? What's the reason?

lindsayad replied:
I like my problems to solve in a reasonable number of nonlinear iterations. If they don't, then I presume there's a problem with my algorithm

recuero replied:
Was not familiar with passing the flag instead of spelling it out through `nl_max_its`. I guess it's just the same.

recuero commented on framework/src/constraints/AutomaticMortarGeneration.C:
```diff
@@ -372,10 +372,15 @@ AutomaticMortarGeneration::buildMortarSegmentMesh()
-    for (MooseIndex(secondary_elem->n_nodes()) n = 0; n < secondary_elem->n_nodes(); ++n)
-      dxyz_dxi += Moose::fe_lagrange_1D_shape_deriv(order, n, xi1) * secondary_elem->point(n);
-    auto normal = Point(dxyz_dxi(1), -dxyz_dxi(0), 0).unit();
+    std::vector<Point> nodal_normals(secondary_elem->n_nodes());
+    for (const auto n : make_range(secondary_elem->n_nodes()))
+      nodal_normals[n] = secondary_node_to_nodal_normal.at(secondary_elem->node_ptr(n));
+    Point normal(0);
+    for (const auto n : make_range(secondary_elem->n_nodes()))
+    {
+      const auto phi = Moose::fe_lagrange_1D_shape(order, n, xi1);
+      normal += phi * nodal_normals[n];
+    }
```
Is this fix to the normal changing any of the existing results? If it isn't, it may only play a role with highly deformed surfaces?

lindsayad replied:
This fix only changes how we determine orientation of the secondary and primary elements (e.g. a kind of "which one is on the left" kind of thing). This should only improve our ability to make that determination...it shouldn't change any of the existing determinations in our tests.

Right, this should only be a factor when we have highly deformed surfaces.

recuero commented on modules/combined/test/tests/gap_heat_transfer_mortar/finite-2d-rz/finite.i:
```diff
@@ -0,0 +1,356 @@
+  [frictionless_normal_lm]
+    order = FIRST
+    block = 'frictionless_secondary_subdomain'
+    scaling = 1e3
```
Any alternative to the individual scaling of the variables for this problem?

lindsayad replied:
We still do not have an automated way of scaling variables that may not have diagonal dependence. I have an open issue for it here: https://github.com/idaholab/moose/issues/16212

recuero reviewed (commented):
👍
